### PR TITLE
sstables: add test of sstables_mutation_reader for missing partition_end

### PR DESCRIPTION
**Pull request by haaawk**
Reproduces #4783
    
Issue was fixed by 9b8ac5ecbc66de574b8ce885de809cbe08d7c16c
    
Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

### CODE REVIEW

**Comment by haaawk** [comment]
@avikivity @glommer @tgrabiec please have a look.

**Comment by avikivity** [comment]
Please adjust the subject line to be self contained. Nobody has any idea what scylla-enterprise#968 is. The commit message should also explain what the problem was, and refer to the patch in master that fixed it.

**Comment by haaawk** [comment]
@avikivity done

File: tests/sstable_mutation_test.cc [hunk 1]
```diff
@@ -817,6 +817,60 @@ SEASTAR_THREAD_TEST_CASE(tombstone_in_tombstone2) {
     }).get();
 }
 
+// Reproducer for enterprise #968
+static schema_ptr buffer_overflow_schema() {
```
Comment (avikivity) [inline tests/sstable_mutation_test.cc]: Please adjust to correct issue

Comment (haaawk) [reply tests/sstable_mutation_test.cc]: Ups. Sorry about that.

Comment (haaawk) [reply tests/sstable_mutation_test.cc]: done.

**Comment by bhalevy** [comment]
With 13f897e3280739334fc04c752a8aa508565487c9 `sstable_mutation_test` fails like this:
```
build/debug/tests/sstable_mutation_test_g -t buffer_overflow --report_level=no --logger=HRF,test_suite:XML,test_suite,jenkins_test.debug.sstable_mutation_test.boost.xml -- -c2 -m2G --overprovisioned --unsafe-bypass-fsync 1 --blocked-reactor-notify-ms 2000000 --collectd 0
system.cloudius has address 10.0.0.4
At work, using machines from /local/home/bhalevy/.distcc/hosts-office
DISTCC_HOSTS=" ran/4 sif/4 loki/4 huginn/4 muninn/4 thor/4 dgccd201/4 system/4  localhost/3"
Using 37 concurrent compilation threads
[1/3] NINJA seastar/build/debug/libseastar.a
ninja: Entering directory `seastar/build/debug'
ninja: no work to do.
Running 1 test case...
Entering test module "tests/sstable_mutation_test.cc"
tests/sstable_mutation_test.cc(0): Entering test case "buffer_overflow"
WARNING: debug mode. Not for benchmarking or production
WARN  2019-08-04 13:25:07,719 seastar - Seastar compiled with default allocator, heap profiler not supported
WARN  2019-08-04 13:25:07,720 seastar - Seastar compiled with default allocator, heap profiler not supported
random-seed=3461980673
==1467==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
Expecting partition start with key {key: pk{000400000004}, token:-2729420104000364805}
Expect {range_tombstone: start={bound: prefix=ckp{000400000002}, kind=incl start}, end={bound: prefix=ckp{000400000002}, kind=incl end}, {tombstone: timestamp=1564914308538769, deletion_time=1564914308}}
Expect ckp{000400000002000400000002}
Expecting partition end
Expecting partition start with key {key: pk{000400000003}, token:9010454139840013625}
Expect {range_tombstone: start={bound: prefix=ckp{000400000001}, kind=incl start}, end={bound: prefix=ckp{000400000001}, kind=incl end}, {tombstone: timestamp=1564914308538769, deletion_time=1564914308}}
Expect ckp{000400000001000400000002}
Expecting partition end
Expecting end of stream
tests/sstable_mutation_test.cc(0): Leaving test case "buffer_overflow"; testing time: 843569us
Leaving test module "tests/sstable_mutation_test.cc"; testing time: 843642us
sstable_mutation_test_g: /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/shared_ptr_debug_helper.hh:63: void seastar::debug_shared_ptr_counter_type::check() const: Assertion `_cpu == std::this_thread::get_id()' failed.
Aborted
```

**Comment by avikivity** [comment]
Unmerged, please fix in debug mode.

File: tests/sstable_mutation_test.cc [hunk 2]
```diff
@@ -817,6 +817,60 @@ SEASTAR_THREAD_TEST_CASE(tombstone_in_tombstone2) {
     }).get();
 }
 
+// Reproducer for #4783
+static schema_ptr buffer_overflow_schema() {
+    static const auto s = [] {
+        schema_builder builder(make_lw_shared(schema(generate_legacy_id("test_ks", "test_tab"), "test_ks", "test_tab",
```
Comment (bhalevy) [inline tests/sstable_mutation_test.cc]: @haaawk make_lw_shared is called by this static lambda.
Could this cause the error we saw:
```
sstable_mutation_test_g: /local/home/bhalevy/dev/scylla/seastar/include/seastar/core/shared_ptr_debug_helper.hh:63: void seastar::debug_shared_ptr_counter_type::check() const: Assertion `_cpu == std::this_thread::get_id()' failed.
```

Comment (haaawk) [reply tests/sstable_mutation_test.cc]: seems that it might be it. Thanks @bhalevy

**Comment by haaawk** [comment]
@avikivity I've pushed fixed version of the test. Please have a look.

**Comment by avikivity** [comment]
What are the changes?

**Comment by haaawk** [comment]
"static thread_local auto s =" instead of "static const auto s ="

**Pull request merged**